Theme Showcase: New component ThemesToolbarGroup

taipeicoder opened the pull request:
#### Proposed Changes

This PR updates the filter toolbar group below the search component as per design spec in 0EogtxvlmxMGcnEEaMVECv-fi-1259%3A37689. More details are written in #69925.

This PR also moves the pricing filter next to the `<ThemesToolbarGroup />` component, its behavior is unchanged from the one in production. To be specific - when users click on "Free" or "Premium", the `<ThemesToolbarGroup />` is set to "All".

| Before | After |
| --- | --- |
| ![Screen Shot 2022-11-16 at 1 46 30 PM](https://user-images.githubusercontent.com/797888/202094057-4a14b99d-374e-4735-a7fc-b514a3ca12fb.png) | ![Screen Shot 2022-11-18 at 11 34 21 AM](https://user-images.githubusercontent.com/797888/202610956-af4deb82-c7d8-4b58-9119-d565f624dcef.png) |

Note that currently, the contrast between the pricing filter component and the background is not good. A separate PR will be created to revamp the Theme Showcase layout aesthetically fit the new components.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Head to the Themes Showcase `/themes/${site_slug}`
  * If using calypso.live, the flag `themes/showcase-i4/search-and-filter` is required.
* Ensure that the filter result set is the same as `wordpress.com/themes/${site_slug}`.
* Ensure that the old filter component works as it used to.
  * Disable the new component via flag, i.e.: `/themes/${site_slug}?flags=-themes/showcase-i4/search-and-filter`
* Ensure that the pricing filter component works as it used to. 
  * Disable the new component via flag, i.e.: `/themes/${site_slug}?flags=-themes/showcase-i4/search-and-filter`

Note that there are 2 filters we deliberately filter out:
1. "Recommended" & "Trending" - Not shown due to being conceptually outdated.
2. "My Themes" - Not shown for simple sites.
3. "Newsletter" - Not shown due to lack of themes.

#### Pre-merge Checklist
<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #69925

taipeicoder committed:
New ThemesToolbarGroup component

taipeicoder committed:
Added config flag for component ThemesToolbarGroup

taipeicoder committed:
UI tuning

fushar commented:
One thing that I noticed is that, after clicking a subject filter (e.g. Blog), then clicking on a theme (which shows the theme details page), if we click the back button, the filter is not preserved. It's pretty annoying. However, the existing behavior in production is also like that, so maybe we can create a new enhancement issue for this 😄

taipeicoder committed:
Added pricing filters

taipeicoder commented:
> One thing that I noticed is that, after clicking a subject filter (e.g. Blog), then clicking on a theme (which shows the theme details page), if we click the back button, the filter is not preserved. It's pretty annoying. However, the existing behavior in production is also like that, so maybe we can create a new enhancement issue for this 😄

That's a good point 🙂 I think we can address this in the next phase where we are changing the theme cards and preview to be the same as onboarding. The theme details page will be gone then!

fushar commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -138,6 +128,49 @@ class ThemeShowcase extends Component {
 	}
 
+	getTabTiers = () => {
+		const { translate } = this.props;
```
I believe `props` is passed from the constructor, so we can use it?

taipeicoder replied:
Do you mean pass `props` as function argument, and we do:
```jsx
getTabTiers = ( { translate } ) => { ... }
```

fushar replied:
That's one way to do it. But I just mean that we either get from `this.props` or pass the `props` from the constructor consistently. Because currently we do:

```
this.tabTiers = this.getTabTiers( props );
this.tabFilters = this.getTabFilters( props );
```

taipeicoder replied:
🤦🏻‍♂️ you're right, I didn't notice that I was already passing props.

fushar commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -138,6 +128,49 @@ class ThemeShowcase extends Component {
+
+	getTabFilters = () => {
+		const { subjects = {}, translate } = this.props;
+		const subjectFilters = Object.fromEntries(
```
Hmm, should we add `order` field to each of the subject filters? Otherwise I'm not sure how the `sort()` would work

```
								<NavTabs>
									{ Object.values( this.tabFilters )
										.sort( ( a, b ) => a.order - b.order )
```

taipeicoder replied:
`order` is only used for the old filter, the new component doesn't need this. We also don't need to add `order` to the "subject" filters since the old filter won't have them (as seen [here](https://github.com/Automattic/wp-calypso/pull/69972/files#diff-5e6c175ac60d9ed48935454e0bcf1861c5d5893c33a08edb946162d2e0dea550R170)).

To simpler words, the "subject" filters don't need `order` since only the new filter component will show them. And the new filter component doesn't require the `order` prop.

fushar replied:
Ah, I misread the code then. Right right. Thanks!

fushar commented:
I think the active pricing filter color is too subtle; I didn't notice the whole thing at first. Has this color scheme been used anywhere else in Calypso? cc @SaxonF 

<img width="273" alt="image" src="https://user-images.githubusercontent.com/1525580/202141398-e054f577-da7f-4b6e-bfe7-b3f03c352919.png">

fushar commented:
> Has this color scheme been used anywhere else in Calypso

Ah, I see that it's coming from an existing component `calypso/components/segmented-control/simplified`...

taipeicoder commented:
> > Has this color scheme been used anywhere else in Calypso
> 
> Ah, I see that it's coming from an existing component `calypso/components/segmented-control/simplified`...

The new layout will have a lighter background (#FDFDFD instead of #F6F7F7) that will make the component background more apparent:
![Screen Shot 2022-11-16 at 5 42 41 PM](https://user-images.githubusercontent.com/797888/202145719-c34d8ebf-973b-4351-a293-c39325926649.png)

fushar commented:
> The new layout

Sorry what do we mean by "new layout"? Will there be followup PR? 😄

taipeicoder committed:
Added criteria for showing the pricing filter

taipeicoder commented:
> > The new layout
> 
> Sorry what do we mean by "new layout"? Will there be followup PR? 😄

Sorry for not being clearer! There is also going to be a slight layout UI revamp to be more similar to the Plugins page. So the Theme Showcase will have its header and page background updated 🙂

| Before | After |
| --- | --- |
| ![Screen Shot 2022-11-16 at 6 21 09 PM](https://user-images.githubusercontent.com/797888/202154727-c88fd54e-01e3-41a2-a744-b1316e7af6cd.png) | ![Screen Shot 2022-11-16 at 6 21 22 PM](https://user-images.githubusercontent.com/797888/202154780-91af1ca3-62bb-4dbd-9c98-44630fbb4401.png) |

You can see the planned order list of tasks here: pbxlJb-2V0-p2

fushar commented:
> There is also going to be a slight layout UI revamp to be more similar to the Plugins page.

Ah, this makes sense now, thanks!

arthur791004 commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -138,6 +129,47 @@ class ThemeShowcase extends Component {
+
+	getTabFilters = ( { subjects = {}, translate } ) => {
+		const subjectFilters = Object.fromEntries(
+			shuffle(
```
Do we need to shuffle the filters all the time? I assume it's a little weird the order is changed when you go into the theme details and back.

taipeicoder replied:
Unfortunately, the entire component is re-mount when coming back from theme details. I've removed the randomization for now, since I couldn't find a clean way to do it without perhaps creating a new function in `client/lib`. Since it's not part of spec, we can do this later. What do you think?

arthur791004 replied:
Sounds good 👍

arthur791004 commented on client/my-sites/themes/themes-toolbar-group/index.tsx:
```diff
@@ -0,0 +1,36 @@
+	onSelect: ( selectedSlug: string | null ) => void;
+}
+
+const ThemeToolbarGroup: React.FC< ThemesToolbarGroupProps > = ( {
```
It should be

```diff
- ThemeToolbarGroup
+ ThemesToolbarGroup
```

taipeicoder replied:
Ah nice catch! 👍

arthur791004 commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -410,39 +461,57 @@ class ThemeShowcase extends Component {
+							<div className="theme__filters">
+								<ThemesToolbarGroup
+									items={ Object.values( this.tabFilters ).filter( ( tabFilter ) =>
+										this.shouldShowTab( tabFilter.key )
```
I guess we're able to filter the `tabFilters` at the beginning and related props changed. Then, we can calculate `selectedKeyIndex` when `onFilterClick` is triggered and get rid of `useEffect` inside `ThemesToolbarGroup`

taipeicoder replied:
Sorry, I don't quite understand the suggestion 😅 Are you suggesting:
1. Instead of using `shouldShowTab` in JSX, we can pre-filter tabs in `getTabFilters()`?
2. Instead of passing `selectedKey`, we save `tabFilterIndex` in `<ThemeShowcase />` state and pass the index to `<ThemesToolbarGroup />`? If that's the case, we would need to either sync both `tabFilter` and `tabFilterIndex` in the component state, or remove `tabFilter` in favor of `tabFilterIndex`. I prefer to keep using `tabFilter` to avoid refactoring for now. What do you think?

Would appreciate further clarification, thank you!

arthur791004 replied:
Sorry for being unclear!

> 1 . Instead of using shouldShowTab in JSX, we can pre-filter tabs in getTabFilters()?

Yes! I think the `isJetpackSite`, `isAtomicSite`, and `siteCanInstallThemes` props won't change when the component re-renders, so we might be able to pre-filter tabs instead of filtering on every render.

> 2. Instead of passing selectedKey, we save tabFilterIndex in <ThemeShowcase /> state and pass the index to <ThemesToolbarGroup />? If that's the case, we would need to either sync both tabFilter and tabFilterIndex in the component state, or remove tabFilter in favor of tabFilterIndex. I prefer to keep using tabFilter to avoid refactoring for now. What do you think?

Yes, what I want is to get rid of `useEffect` to calculate `activeIndex`. But now I think `useMemo` is enough 🙂
See https://beta.reactjs.org/learn/you-might-not-need-an-effect#caching-expensive-calculations

arthur791004 commented on client/my-sites/themes/theme-showcase.scss:
```diff
@@ -77,6 +77,54 @@
+	justify-content: center;
+	padding: 16px 0 24px;
+
+	@include breakpoint-deprecated( ">660px" ) {
```
Could we change to use `break-medium` or `break-medium` as the mixin is deprecated?

taipeicoder replied:
I updated to `break-small` (600px) instead, since it's also referenced [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/themes/theme-showcase.jsx#L144). I think in the future, we will need to update all the breakpoints `>660px` to `break-small`.

arthur791004 commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -410,39 +461,57 @@ class ThemeShowcase extends Component {
+									}
+								/>
+								{ premiumThemesEnabled && ! isMultisite && (
+									<SimplifiedSegmentedControl
```
Do we need to show it on the new version? I cannot see it on the Figma 🤔

taipeicoder replied:
The spec can be found here: 0EogtxvlmxMGcnEEaMVECv-fi-1492%3A52007&t=K1fZL0gHpcnkIoOG-0
The PT also has a screenshot of the design: pbxlJb-2V0-p2

taipeicoder committed:
Address feedback

fushar commented:
@taipeicoder I just found out that we could not combine the subject and pricing filter together. For example, it is not currently possible to search for Portfolio AND premium themes. Is this intended?

fushar commented:
☝️ whereas, it is possible to do that in production: https://wordpress.com/themes/premium/filter/subject:portfolio/testfushar36.wordpress.com

fushar reviewed (changes requested):
Blocking until we resolve the concern around combining subject and pricing filters

taipeicoder committed:
Address feedback

taipeicoder commented:
> ☝️ whereas, it is possible to do that in production: https://wordpress.com/themes/premium/filter/subject:portfolio/testfushar36.wordpress.com

This is what I observed (please confirm if it's your experience as well):
1. Setting the pricing to "Premium", then in the search component, select Subject → Portfolio.
2. Setting the pricing to "Premium", then clicking on the filter "Portfolio" would result in the pricing being set back to "All".
3. Setting the filter to "Portfolio", then setting the pricing to "All" would result in the filter being set back to "All".

Point 2 & 3 is carried over from the existing component where if any pricing other than "All" is set, then the subject filters will go back to "All". And vice-versa, where if any subject other than "All" is set, then the pricing will go back to "All". Not sure if this is what you meant 🤔

I think ultimately we can improve the logic, as long as it doesn't overcomplicate the code because for the moment we need to take backwards compatibility into consideration 🥲 Or we can just update the logic altogether.

taipeicoder commented:
> > ☝️ whereas, it is possible to do that in production: https://wordpress.com/themes/premium/filter/subject:portfolio/testfushar36.wordpress.com
> 
> This is what I observed (please confirm if it's your experience as well):
> 
> 1. Setting the pricing to "Premium", then in the search component, select Subject → Portfolio.
> 2. Setting the pricing to "Premium", then clicking on the filter "Portfolio" would result in the pricing being set back to "All".
> 3. Setting the filter to "Portfolio", then setting the pricing to "All" would result in the filter being set back to "All".
> 
> Point 2 & 3 is carried over from the existing component where if any pricing other than "All" is set, then the subject filters will go back to "All". And vice-versa, where if any subject other than "All" is set, then the pricing will go back to "All". Not sure if this is what you meant 🤔
> 
> I think ultimately we can improve the logic, as long as it doesn't overcomplicate the code because for the moment we need to take backwards compatibility into consideration 🥲 Or we can just update the logic altogether.

Following up on this: The reason that the filters "snap" back to "All" is that "Recommended" and "Trending" currently [do](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/themes/actions/recommended-themes.js) [not](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/themes/actions/trending-themes.js) support query options.

fushar commented:
@taipeicoder For example, I am not able to make both the "Portfolio" filter and the "Premium" filter active at the same time.

taipeicoder commented:
> > 1. Setting the pricing to "Premium", then in the search component, select Subject → Portfolio.
> 
> I did the following: set the pricing to "Premium" and then click "Portfolio". It is not possible.
> 
> Also, is it equivalent to "Subject -> Portfolio"? Should searching for that subject activate the "Portfolio" filter (and vice versa?)

> Setting the pricing to "Premium", then in the **search component**, select Subject → Portfolio.
I meant this 😅:
![Screen Shot 2022-11-18 at 8 45 13 AM](https://user-images.githubusercontent.com/797888/202590622-a1a8887f-4bfc-42ae-a690-96d86194bc86.png)

> Also, is it equivalent to "Subject -> Portfolio"? Should searching for that subject activate the "Portfolio" filter (and vice versa?)
I don't we need it since it's going to be a rabbit hole trying to sync all those 3 components (search + category filter + pricing filter) 🥲 Currently, the logic of "search" only applies to the "All" tab makes sense to me, albeit a bit less powerful. This would mean that we would need to clear the search input when users change tabs, which is not the production behavior. Maybe we can create a search UX improvement task?

>For example, I am not able to make both the "Portfolio" filter and the "Premium" filter active at the same time. 
As I explained in the [previous comment](https://github.com/Automattic/wp-calypso/pull/69972#issuecomment-1318593583), the behavior in production is that pricing filters and category filters "snap" back to "All" whenever one of them is not "All". This is because "Recommended" and "Trending" doesn't have filter/tier/search implemented.

However, the new ones we added ("Blog", "Portfolio", etc) support filter/tier/search, which means that they don't have to snap back to "All". That said, it would create an inconsistent UX where only "Recommended" and "Trending" would snap while others won't. I think ideally we follow up by looking if it's possible to make "Recommended" and "Trending" query-able and then make all filters to not snap anymore. Sorry if I'm not explaining this very well.!

fushar commented:
> However, the new ones we added ("Blog", "Portfolio", etc) support filter/tier/search, which means that they don't have to snap back to "All". That said, it would create an inconsistent UX where only "Recommended" and "Trending" would snap while others won't.

I see now what you mean! Yeah, that's sad. But I do believe that we should be able to combine the filters; the current behavior in this PR is confusing. We can follow-up in a later PR I guess 😄

taipeicoder committed:
Address feedback

taipeicoder commented:
> > However, the new ones we added ("Blog", "Portfolio", etc) support filter/tier/search, which means that they don't have to snap back to "All". That said, it would create an inconsistent UX where only "Recommended" and "Trending" would snap while others won't.
> 
> I see now what you mean! Yeah, that's sad. But I do believe that we should be able to combine the filters; the current behavior in this PR is confusing. We can follow-up in a later PR I guess 😄

+1. I've updated the behavior, since we are tentatively removing the filters "Recommended" and "Trending" for the new experience. We are also updating the URL when selecting filters!

taipeicoder committed:
Persist search component filter+search when switching subject filters

taipeicoder committed:
Update selected filter according to search component

taipeicoder committed:
Fixed undefined data handling

fushar reviewed (approved):
Nice work!

One small issue left: when I'm filtering for `subject:podcast` in the search box, the Podcast filter is automatically selected. But it is not the other way around: If I click on Podcast, the search box is not automatically filled with `subject:podcast`.

Not sure if we should do that. We can follow up in future PR I think.

taipeicoder commented:
> Nice work!
> 
> One small issue left: when I'm filtering for `subject:podcast` in the search box, the Podcast filter is automatically selected. But it is not the other way around: If I click on Podcast, the search box is not automatically filled with `subject:podcast`.
> 
> Not sure if we should do that. We can follow up in future PR I think.

Follow-up issue: https://github.com/Automattic/wp-calypso/issues/70162

taipeicoder committed:
Fixed undefined data handling
